### ISSUE TITLE
Reproduce wmt16 en-de results in latent depth

### ISSUE THREAD

**Issue by lemmonation** [pytorch/fairseq#2870]
@xianxl 
Firstly, thanks a lot for proposing the novel work and releasing the code!
I have some questions about the implementation details, and it would be nice if you can give some advice.

1) Could you provide the training script of the wmt16 en-de task? Or could you point out the difference of hyper-parameter settings between this setting and the multilingual setting?

2) I'm confused about the "sparsity_weight" and "shared_weight". It seems that the kl loss is weighted by "sparsity_weight" while the sparsity loss is weighted by "shared_weight".

3) In the bilingual setting, is it still necessary to add the ["share_loss"](https://github.com/pytorch/fairseq/blob/master/examples/latent_depth/latent_depth_src/loss/latent_depth.py#L82)? It seems that the sparsity loss as well as the share loss are both weighted by shared_weight, therefore cannot be omitted exclusively.

4) I find that there will be an OOM error with update-freq>1. I guess it is caused by [retain_graph=True](retain_graph=True), am I right? And in the bilingual setting, can I omit this setting and write line [122-141](https://github.com/pytorch/fairseq/blob/master/examples/latent_depth/latent_depth_src/multilingual_translation_latent_depth.py#L122-L141) into function _per_lang_pair_train_loss(), just treat the sparsity loss the same as kl loss?

**Comment by AIikai** [comment]
> @xianxl
> Firstly, thanks a lot for proposing the novel work and releasing the code!
> I have some questions about the implementation details, and it would be nice if you can give some advice.
> 
> 1. Could you provide the training script of the wmt16 en-de task? Or could you point out the difference of hyper-parameter settings between this setting and the multilingual setting?
> 2. I'm confused about the "sparsity_weight" and "shared_weight". It seems that the kl loss is weighted by "sparsity_weight" while the sparsity loss is weighted by "shared_weight".
> 3. In the bilingual setting, is it still necessary to add the ["share_loss"](https://github.com/pytorch/fairseq/blob/master/examples/latent_depth/latent_depth_src/loss/latent_depth.py#L82)? It seems that the sparsity loss as well as the share loss are both weighted by shared_weight, therefore cannot be omitted exclusively.
> 4. I find that there will be an OOM error with update-freq>1. I guess it is caused by [retain_graph=True](retain_graph=True), am I right? And in the bilingual setting, can I omit this setting and write line [122-141](https://github.com/pytorch/fairseq/blob/master/examples/latent_depth/latent_depth_src/multilingual_translation_latent_depth.py#L122-L141) into function _per_lang_pair_train_loss(), just treat the sparsity loss the same as kl loss?

Excuse me, are you clear about these problems? Especially the first one.
